docs/design/specs/runtime-async.md

```diff
@@ -0,0 +1,68 @@
+
+This document is a draft of changes to ECMA-335 for the "runtime async" feature. When the feature is officially supported, it can be merged into the final ECMA-335 augments document.
+
+# Runtime-async
+
+Async is currently a feature implemented by various .NET languages as a compiler rewrite to support methods that can "yield" control back to their caller at specific "suspension" points. While effective, it's believed that implementation directly in the .NET runtime can provide improvements, especially in performance.
+
+## Spec modifications
+
+These are proposed modifications to the ECMA-335 specification for runtime-async.
+
+### I.8.4.5 Sync and Async Methods
+
+Methods may be either 'sync' or 'async'. Async method definitions are methods with a return type of `System.Threading.Task`, `System.Threading.ValueTask`, `System.Threading.Task<T>`, or `System.Threading.ValueTask<T>`  attributed with `[RuntimeAsyncMethodAttribute]`. Inside async method bodies, async methods are also invokable by a special signature encoding, described in [### I.8.6.1.5 Method signatures].
```

Comment:
We should explicitly spell out the shape of `RuntimeAsyncMethodAttribute` in this proposal, including containing namespace and constructors. Also, do we intend to define that attribute in the same assembly as `System.Object`? If so, we can use its presence to drive compiler behavior around the new form; the alternative is to use a `RuntimeFlags` flag, as we've done in the past with other new runtime features. We could also just always use `RuntimeFlags` to drive it, but we do need to specify whether the runtime will honor _any_ `RuntimeAsyncMethodAttribute`, or just the one it defines.

Reply:
I also think we should put `CompilerFeatureRequiredAttribute` on the `RuntimeAsyncMethodAttribute` type, so that compilers who don't understand the feature block its use. We probably don't want users slapping this on their own methods.

Reply:
Included the namespace name and mention of the assembly requirements. Didn't spell out the constructor requirements, but the example usage shows the empty constructor.

We could add stuff like AttributeUsage and CompilerFeatureRequiredAttribute, but they don't have any actual impact on ECMA so it's debatable whether this is the right place, or if that should live in a compiler spec.